feat: add a `command_not_found` hook

stevenxxiu opened the pull request:
# Description
Add a `command_not_found` function to `$env.config.hooks`. If this function outputs a string, then it's included in the `help`.

An example hook on *Arch Linux*, to find packages that contain the binary, looks like:

```nushell
let-env config = {
  # ...
  hooks: {
    command_not_found: {
      |cmd_name| (
        try {
          let pkgs = (pkgfile --binaries --verbose $cmd_name)
          (
            $"(ansi $env.config.color_config.shape_external)($cmd_name)(ansi reset) " +
            $"may be found in the following packages:\n($pkgs)"
          )
        } catch {
          null
        }
      )
    }
    # ...
```

# User-Facing Changes
- Add a `command_not_found` function to `$env.config.hooks`.

# Tests + Formatting

Don't forget to add tests that cover your changes.

Make sure you've run and fixed any issues with these commands:

- `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- `cargo clippy --workspace -- -D warnings -D clippy::unwrap_used -A clippy::needless_collect` to check that you're using the standard code style
- `cargo test --workspace` to check that all tests pass

# After Submitting

If your PR had any user-facing changes, update [the documentation](https://github.com/nushell/nushell.github.io) after the PR is merged, if necessary. This will help us keep the docs up to date.

sholderbach commented on crates/nu-command/Cargo.toml:
```diff
@@ -14,6 +14,7 @@ bench = false
 
 [dependencies]
 nu-ansi-term = "0.46.0"
+nu-cli = { path = "../nu-cli", version = "0.76.1" }
```
Mhh that introduces a circular dependency situation as `nu-cli` itself depends on `nu-command`.

stevenxxiu replied:
Hmm I'm just using `eval_hook` from there. Where do you think this should be moved?

stevenxxiu replied:
I moved this in a new commit.

fdncred commented:
Thanks for your continued contributions to nushell. They are sincerely appreciated.

However :) I'm not a super-fan of this PR. I really don't want to run some external package manager tool every time I try to run an external executable. I see that it's initialized to `null` but it really seems like that would be a total drag on the system waiting for a few seconds with every execution.

Am I missing the point here? Is this something people really need? And is this common in other shells like bash/fish/zsh?

Maybe one way of working around this is to have the ability for people to add a custom hook to do whatever they want and run it whenever they want (limited to some set of run-it-at-these-times list). I'm not really sure what that looks like or even if it's possible.

stevenxxiu commented:
Thanks for your appreciation.

I believe the hook is only run when an executable or command isn't found. That shouldn't happen too often.

In the above example I'm pretty sure that `pkgfile` only runs offline and syncs the repo once a while. It's pretty fast. It has hooks for all of *Bash*/*Fish*/*Zsh* actually. In the years of using *Zsh* I never noticed a slowdown. This hook only ran when either I had a typo, or I thought I installed a package but didn't.

sholderbach commented:
Thanks for resolving this dependency problem.

In general you have my vote for adding a hook to deal with the command not found case.

Two concerns that need to be addressed (and checked how this PR fairs in this regard):

1. hooks should only affect the interactive experience and not affect the runtime behavior of the scripting use of nushell. (that was the reason why the evaluation logic lives in `nu-cli`)
2. the command not found hook must under no circumstances mess with how name resolution behaves. it should only take action if you otherwise would get an error message. 
    - name resolution is a complicated beast already
    - scripts should not break because one config contains a command not found hook and the other doesn't
3. the output of the command not found hook evaluation is not interfering with the already failed pipeline

stevenxxiu commented:
Thanks for that. I addressed your concerns:

> 1. hooks should only affect the interactive experience and not affect the runtime behavior of the scripting use of nushell. (that was the reason why the evaluation logic lives in `nu-cli`)

Now I test for `if engine_state.is_interactive {` in `crates/nu-command/src/hook.rs`. I checked to see if I can just print the error instead in `crates/nu-cli/src/repl.rs`, but this doesn't seem easy. It just runs `eval_source()`. This hides the details of what's going on.

> 2. the command not found hook must under no circumstances mess with how name resolution behaves. it should only take action if you otherwise would get an error message.
>    
>    * name resolution is a complicated beast already
>    * scripts should not break because one config contains a command not found hook and the other doesn't

This should be always the case. In `crates/nu-command/src/system/run_external.rs`, the hook only runs when the  `ExternalCommand.run_with_input()` function returns `Err(ShellError::ExternalCommand { ... })`.

> 3. the output of the command not found hook evaluation is not interfering with the already failed pipeline

I updated the code so the code only ever does something if the `command_not_found` hook returns a string.

stevenxxiu commented:
No new changes here. Just rebased to `main`.

jntrnr commented on crates/nu-command/src/system/run_external.rs:
```diff
@@ -324,9 +325,34 @@ impl ExternalCommand {
+                            if let Some(hook) = engine_state.config.hooks.command_not_found.clone()
+                            {
+                                if let Ok(PipelineData::Value(Value::String { val, .. }, ..)) =
+                                    eval_hook(
```
Would it be running the `command_not_found` hook instead of returning an error if they set it?

I'm imagining something like a distro my add this hook to go find a command and help you install it. Once installed, the hook returns. I think we'd still show an error in that case?

stevenxxiu replied:
It would still return an error. I can see some issues with helping people install packages from errors.

- A single command can be provided by multiple packages.
- Every error about command typos giving a package install interface, feels a bit annoying.

It also doesn't apply to the *Arch Linux* case, as the only way to install things is to manually run *Pacman*.

jntrnr replied:
Interesting - can you describe the use case where you want to run a command but still give the error? I can't think of one but maybe you have one in mind

stevenxxiu replied:
Well say I have some pretty long command/script I enter, and within that script there's some typo. In that case I think it's pretty useful to see where the typo is. It's better than the output just telling me where to find, or that this command with a typo in it can't be found.

jntrnr replied:
Oh okay, got it. Yeah now that I see how it works together that makes sense.

flying-sheep commented:
Thank you, that’s great! We could add an `if ($pkgs | is-empty) { null } else { ( $"..." + $"..." ) }` to your code and add it as an example to the docs.

stevenxxiu commented:
No worries. It was one of the things I wanted to add when I switched to *Nushell*.

I'm not too sure what's going on in the example you gave tbh. Maybe I can just add the *Arch Linux* example? It does work so is pretty practical.

flying-sheep commented:
What I’m saying is that this example needs a small improvement:

```diff
 let-env config = {
   # ...
   hooks: {
     command_not_found: {
       |cmd_name| (
         try {
           let pkgs = (pkgfile --binaries --verbose $cmd_name)
-          (
+          if ($pkgs | is-empty) {
+            null
+          } else { (
             $"(ansi $env.config.color_config.shape_external)($cmd_name)(ansi reset) " +
             $"may be found in the following packages:\n($pkgs)"
-           )
+           ) }
         } catch {
           null
         }
       )
     }
     # ...
 }
```

fdncred commented:
@stevenxxiu A good example is great for documentation. Please just be specific about which OS/Distro this applies to. e.g. I assume that this would fail in Windows, and probably MacOS.
